### PR TITLE
Update ua_binary.py

### PR DESCRIPTION
**Pull request by CuriousCrook**
Fix for issue #1127 

Ran 410 tests in 75.357s
OK

All tests ran file and the connection to the Softing Server now also works for me.

### CODE REVIEW

**Comment by AndreasHeine** [comment]
those port errors... maybe we should modify the test to try/except for next free port.
just commit a blank line and the test will be run again!

**Comment by CuriousCrook** [comment]
I didn't understand the problem, but it's working now ;-)

File: opcua/ua/ua_binary.py [hunk 1]
```diff
@@ -503,6 +503,9 @@ def struct_from_binary(objtype, data):
             if not test_bit(val, idx):
                 continue
         val = from_binary(uatype, data)
+        if val is None and uatype[:6] == 'ListOf':
+            val = []
+
```
Comment (oroulet) [inline opcua/ua/ua_binary.py:508]: What is the issue exactly?the UA protocols allows to return Null values which are best translated to None in Python. So the original code looked correct

Comment (oroulet) [inline opcua/ua/ua_binary.py:508]: What is the issue exactly?the UA protocols allows to return Null values which are best translated to None in Python. So the original code looked correct

**Comment by AndreasHeine** [comment]
@CuriousCrook can you track down where the doscovery urls got overwritten? its got initialized as an empty list!?

File: opcua/ua/ua_binary.py (hunk 1 above)
Comment (CuriousCrook) [reply opcua/ua/ua_binary.py:508]: Because the empty lists are iterated in the more high level modules. A non iterable NoneType
leads to an error there. See issue #1127. This error occurred when trying to communicate with a Softing server.

For example:
File "C:_repo\opcua-client-gui-master\venv\lib\site-packages\opcua\tools.py", line 423, in application_to_strings
for url in app.DiscoveryUrls:
TypeError: 'NoneType' object is not iterable

Instead of checking the lists against None everywhere I chose to fix this issue on its roots.

**Comment by CuriousCrook** [comment]
> @CuriousCrook can you track down where the doscovery urls got overwritten? its got initialized as an empty list!?

See above. 
It is overwritten in opcua / ua / ua_binary.py.

In my opinion it is initialized 

In my opinion the list DiscoveryUrls is initialized in a meaningful way. Then it is overwritten by the server response with None.

You can just call application_to_strings(app): in python-opcua/opcua/tools.py while connected to a server without DiscoveryUrls. to reproduce this error. The Softing server seems to return "null", which seems to be valid by the opc ua specification.

I would also like to write a Test, but I haven't found out how to initialize the DiscoveryUrls in the python-opcua server with None.

**Comment by CuriousCrook** [comment]
Let's get the facts together, then we can make a fact-based decision:

Solution according to the Pull-Request
pro:
- Uniform representation of empty lists (initialization status and after a response from server)
- Universal solution for all lists
- List do not have to be checked against None before using (seems more pythonic to me)
- Change only necessary in one place in the source code

contra:
- May slow down communication. (just a guess. Can we measure whether this change is actually causing a relevant slowdown?)

Do you know any other advantages or disadvantages?

**Comment by oroulet** [comment]
but if you change the protocol to return [] everytime we receiver NULL then you may get screwed if one days at one place it is necessary to receiver NULL/None because it has a special meaning

**Comment by oroulet** [comment]
unless someone demonstrate that empty lists cannot be null in UA spec we cannot merge this

**Comment by CuriousCrook** [comment]
This solution was not accepted.
